tools/snapshot_bot_remote_display/snapshot_bot_remote_display.cc

```diff
@@ -0,0 +1,436 @@
+                                                                            1, imageWidth - numScanColumns, imageWidth);
+                    rotatorRight.rotate(
+                            [imageSize, numScanColumns, &differencer, this]
+                            (const cv::Mat &fr, const cv::Mat &, size_t i)
+                            {
+                                // Calculate image difference
+                                auto diffIter = differencer(fr, m_BestSnapshot, m_SnapshotDifferenceScratch);
+
+                                float sumDifference = std::accumulate(diffIter, diffIter + imageSize, 0.0f);
+
+                                // Store mean difference in RIDF
+                                m_RIDF[(2 * numScanColumns) - 1 - i] = Navigation::AbsDiff::mean(sumDifference, imageSize);
+                             });
+
+                    // Get range of
+                    const float minRIDF = *std::min_element(m_RIDF.cbegin(), m_RIDF.cend());
+                    const float maxRIDF = *std::max_element(m_RIDF.cbegin(), m_RIDF.cend());
+
+                    // Get RIDF axis rect and calculate scale factors
+                    const auto &ridfAxisRect = m_Config.getTestingRIDFAxisRect();
+                    const float ridfXScale = (float)ridfAxisRect.width / (float)(numScanColumns * 2);
+                    const float ridfYScale = (float)(ridfAxisRect.height - m_Config.getTestingRIDFLineThickness()) / (maxRIDF - minRIDF);
+
+                    // Get axis ROI
+                    cv::Mat roiRIDF(m_OutputImage, ridfAxisRect);
+
+                    // Clear axis
+                    roiRIDF.setTo(m_Config.getTestingRIDFBackgroundColour());
+
+                    // Draw RIDF line graph
+                    cv::Point lastPoint(0, ridfAxisRect.height - (int)std::round((m_RIDF[0] - minRIDF) * ridfYScale));
+                    for(int i = 1; i < (2 * numScanColumns); i++) {
+                        const cv::Point point(std::round(i * ridfXScale),
+                                              ridfAxisRect.height - (int)std::round((m_RIDF[i] - minRIDF) * ridfYScale));
+                        cv::line(roiRIDF, lastPoint, point, m_Config.getTestingRIDFLineColour(), m_Config.getTestingRIDFLineThickness(), cv::LINE_AA);
+                        lastPoint = point;
+                    }
+
+                    // Get index of min RIDF element and hence calculate pixels to roll snapshot
+                    // **YUCK** this information was clearly present deep within Navigation::InSilicoRotater
+                    const int minRIDFIdx = (int)std::distance(m_RIDF.cbegin(), std::min_element(m_RIDF.cbegin(), m_RIDF.cend()));
+                    const int rollPixels = (minRIDFIdx < numScanColumns) ? (numScanColumns - 1 - minRIDFIdx) : (imageWidth + numScanColumns - 1 - minRIDFIdx);
+                    BOB_ASSERT(rollPixels >= 0);
+                    BOB_ASSERT(rollPixels < imageWidth);
```

Comment:
Out of interest, is there a reason that you didn't use ``PerfectMemoryRotater`` for this? (Other than the sprawling confusing interface :wink:).

``PerfectMemory::getHeading()`` returns a tuple with the best-matching snapshot as the second element (if you're using ``BestMatchingSnapshot``).

Reply:
Hmm, good point. Using a perfect memory, adding the single image and then clearing it would probably be nicer than this nightmare